File: java/datastructures/MinHeap.java
// Min Heap implementation in Java
// Basic heap operations: insert, extractMin, peek, isEmpty, size, clear, and toArray methods

import java.util.*;

public class MinHeap<T extends Comparable<T>> {
    private List<T> heap;

    public MinHeap() {
        this.heap = new ArrayList<>();
    }

    public void insert(T value) {
        heap.add(value);
        bubbleUp();
    }

    private void bubbleUp() {
        int index = heap.size() - 1;
        T element = heap.get(index);

        while (index > 0) {
            int parentIndex = (index - 1) / 2;
            T parent = heap.get(parentIndex);

            if (element.compareTo(parent) >= 0) break;

            // Swap with parent
            heap.set(parentIndex, element);
            heap.set(index, parent);
            index = parentIndex;
        }
    }

    public T extractMin() {
        if (isEmpty()) {
            return null;
        }

        T min = heap.get(0);
        T last = heap.remove(heap.size() - 1);

        if (heap.size() > 0) {
            heap.set(0, last);
            bubbleDown();
        }

        return min;
    }

    private void bubbleDown() {
        int index = 0;
        T element = heap.get(0);

        while (true) {
            int leftIndex = 2 * index + 1;
            int rightIndex = 2 * index + 2;
            int smallest = index;

            // Find smallest child
            if (leftIndex < heap.size() && heap.get(leftIndex).compareTo(heap.get(smallest)) < 0) {
                smallest = leftIndex;
            }
            if (rightIndex < heap.size() && heap.get(rightIndex).compareTo(heap.get(smallest)) < 0) {
                smallest = rightIndex;
            }

            if (smallest == index) break;

            // Swap with smallest child
            heap.set(index, heap.get(smallest));
            heap.set(smallest, element);
            index = smallest;
        }
    }

    public T peek() {
        return isEmpty() ? null : heap.get(0);
    }

    public boolean isEmpty() {
        return heap.size() == 0;
    }

    public int size() {
        return heap.size();
    }

    public void clear() {
        heap.clear();
    }

    public List<T> toArray() {
        return new ArrayList<>(heap);
    }

    public void display() {
        System.out.println("Heap: " + heap);
    }
} 

File: javascript/datastructure/min-heap.js
// Simple Min Heap implementation in JavaScript
// Basic heap operations: insert, extractMin, peek, isEmpty, size, clear, and toArray methods

class MinHeap {
    constructor() {
        this.heap = [];
    }

    // Insert new element
    insert(value) {
        this.heap.push(value);
        this.bubbleUp();
    }

    // Move element up to maintain heap property
    bubbleUp() {
        let index = this.heap.length - 1;
        const element = this.heap[index];

        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            const parent = this.heap[parentIndex];

            if (element >= parent) break;

            // Swap with parent
            this.heap[parentIndex] = element;
            this.heap[index] = parent;
            index = parentIndex;
        }
    }

    // Remove and return minimum element
    extractMin() {
        if (this.isEmpty()) {
            return null;
        }

        const min = this.heap[0];
        const last = this.heap.pop();

        if (this.heap.length > 0) {
            this.heap[0] = last;
            this.bubbleDown();
        }

        return min;
    }

    // Move element down to maintain heap property
    bubbleDown() {
        let index = 0;
        const element = this.heap[0];

        while (true) {
            const leftIndex = 2 * index + 1;
            const rightIndex = 2 * index + 2;
            let smallest = index;

            // Find smallest child
            if (leftIndex < this.heap.length && this.heap[leftIndex] < this.heap[smallest]) {
                smallest = leftIndex;
            }
            if (rightIndex < this.heap.length && this.heap[rightIndex] < this.heap[smallest]) {
                smallest = rightIndex;
            }

            if (smallest === index) break;

            // Swap with smallest child
            this.heap[index] = this.heap[smallest];
            this.heap[smallest] = element;
            index = smallest;
        }
    }

    // Get minimum element without removing
    peek() {
        return this.isEmpty() ? null : this.heap[0];
    }

    // Check if heap is empty
    isEmpty() {
        return this.heap.length === 0;
    }

    // Get heap size
    size() {
        return this.heap.length;
    }

    // Clear the heap
    clear() {
        this.heap = [];
    }

    // Convert heap to array
    toArray() {
        return [...this.heap];
    }

    // Display heap
    display() {
        console.log("Heap:", this.heap);
    }
}

module.exports = { MinHeap }; 